src/_pytest/_code/code.py

```diff
@@ -1038,20 +1038,27 @@ def __init__(
         self.style = style
 
     def toterminal(self, tw: TerminalWriter) -> None:
+        error_prefix = "E   "
+        src = "\n".join(x for x in self.lines if not x.startswith(error_prefix))
+        error_lines = (x for x in self.lines if x.startswith(error_prefix))
```

Comment:
Highlighting should be done in a single block, with the source only.
Otherwise it does not work with e.g.:
```python
def test():
    print("""
    """); assert 0
```
You might want to check the pdbpp source I've linked you to already, which is similar in that regard (since it removes prefixes added by pdb also).

In general I think this should be factored out into a separate (private) method (using `_highlight` theren then), and then just using `tw.write` with the result.  Makes it also easier to unit test then.

Reply:
> You might want to check the pdbpp source I've linked you to already

Sorry it seems I've missed it (both here and in `py`'s PR). Would you mind linking it again?

> Highlighting should be done in a single block, with the source only.

Isn't it what's being done here? Note that `src` is the whole source string.

Here's what your example looks to me:

![image](https://user-images.githubusercontent.com/1085180/73699545-da66ee80-46c2-11ea-8e62-010731bb6889.png)

Reply:
> In general I think this should be factored out into a separate (private) method (using _highlight theren then)

Great idea! I just disagree a bit about this function also doing highlight: I like the idea of other parts of the terminal writing highlighted source code. I will push in a bit a WIP of what I mean.

Reply:
Actually I'm not so sure anymore of what I came up with; I wanted something to help extracting the "parsing" logic of the traceback. I'm just pushing to have the commit online.

Reply:
">" is highlighted as a comment.

See https://github.com/pdbpp/pdbpp/blob/cb56e1b6126bb87597684d01af6f18c873be85ee/src/pdbpp.py#L1071-L1123 - not sure if it helps, since it does much more.

Reply:
Thanks, will take a look later. :+

Reply:
Indeed that does much more than we need now it seems, thanks for the link thought.